### DISCUSSION TITLE
Provide & Inject for Prop Drilling Like Vue Components

### DISCUSSION THREAD

**Discussion by SamuNatsu** [withastro/roadmap#1084] [Proposal]
### Body

# Background & Motivation

It is annoying that I should pass the props again and again from parent to their children and children's children using Astro components.

So some props drilling ways may help, Vue's [Provide/Inject API](https://vuejs.org/guide/components/provide-inject.html) is a great example.

It is not like [contexts](https://github.com/withastro/roadmap/discussions/445), so it would not have circular references, injects would only accept the values provided by ancestors.

For those who provides the same key, just overwrite the ancestor's value, or provide an option to deceide whether to overwrite it. Maybe some errors or warnings should be throwed when the overwrite occurs.

# Example

In ancestor:

```ts
provide(/* key */ 'message', /* value */ 'hello');
provide(/* key */ 'message', /* value */ 'world', /* Options */ { overwite: true });
```

In descendants:

```ts
const message = inject(/* key */ 'message');
```

**Comment by florian-lefebvre** [comment]
There's a community package https://github.com/astro-community/context

**Reply by BeanWei** [reply]
It is only valid for static output mode.

**Reply by florian-lefebvre** [reply]
I didn't know, TIL!
